test/e2e/apimachinery/crd_watch.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apimachinery
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apiextensions-apiserver/test/integration/testserver"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	watchCRNameA = "name1"
+	watchCRNameB = "name2"
+)
+
+var _ = SIGDescribe("CustomResourceDefinition Watch", func() {
+
+	f := framework.NewDefaultFramework("crd-watch")
+
+	Context("CustomResourceDefinition Watch", func() {
```

Comment:
There was the native-resource watch test some days ago. Is it in the same package? Do we share code with it?

Reply:
Yes, they are in the same package. I've used functions `expectEvent`, `expectNoEvent` and `waitForEvent` from that PR, but because CRs are of type Unstructured and not runtime.Object I can't use them directly. So I've copied those functions in this file. I should probably rename them to something else to avoid name conflict. Please let me know if there's a better way to do this.

Reply:
unstructured.Unstructured is a runtime.Object.

Reply:
Ah, cool. I'll just use the functions directly in that case.

Reply:
I've removed the functions from this file, and instead, I'm using them from watch.go (which is yet to be merged).